File: go/words/tests/words_test.go
package tests

import (
	"testing"

	. "example.com/words/words"
)

func TestReverseString(t *testing.T) {
	cases := []struct {
		s        string
		expected string
	}{
		{"hello", "olleh"},
		{"a", "a"},
		{"", ""},
	}
	for _, c := range cases {
		result := ReverseString(c.s)
		if result != c.expected {
			t.Errorf("ReverseString(%s) = %s; want %s", c.s, result, c.expected)
		}
	}
}

func TestIsPalindrome(t *testing.T) {
	cases := []struct {
		s        string
		expected bool
	}{
		{"race car", true},
		{"level", true},
		{"hello", false},
		{"a", true},
		{"", true},
	}
	for _, c := range cases {
		result := IsPalindrome(c.s)
		if result != c.expected {
			t.Errorf("IsPalindrome(%s) = %t; want %t", c.s, result, c.expected)
		}
	}
}

func TestKMPSearch(t *testing.T) {
	cases := []struct {
		sub      string
		string   string
		expected bool
	}{
		{"test", "this is a test", true},
		{"not", "this is a test", false},
		{"", "any string", true},
		{"abc", "abc", true},
		{"abc", "ab", false},
	}
	for _, c := range cases {
		result := KMPSearch(c.sub, c.string)
		if result != c.expected {
			t.Errorf("KMPSearch(%s, %s) = %t; want %t", c.sub, c.string, result, c.expected)
		}
	}
}

func TestLCS(t *testing.T) {
	cases := []struct {
		str1     string
		str2     string
		expected int
	}{
		{"AGGTAB", "GXTXAYB", 4},
		{"ABC", "ABC", 3},
		{"ABC", "DEF", 0},
		{"", "ABC", 0},
		{"ABC", "", 0},
	}
	for _, c := range cases {
		result := LCS(c.str1, c.str2)
		if result != c.expected {
			t.Errorf("LCS(%s, %s) = %d; want %d", c.str1, c.str2, result, c.expected)
		}
	}
}

File: rust/words/tests/words_test.rs
use words::words::*;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_reverse_string() {
        assert_eq!(reverse_string("hello"), "olleh");
        assert_eq!(reverse_string("a"), "a");
        assert_eq!(reverse_string(""), "");
    }

    #[test]
    fn test_is_palindrome() {
        assert_eq!(is_palindrome("race car"), true);
        assert_eq!(is_palindrome("level"), true);
        assert_eq!(is_palindrome("hello"), false);
        assert_eq!(is_palindrome("a"), true);
        assert_eq!(is_palindrome(""), true);
    }

    #[test]
    fn test_is_substring() {
        assert_eq!(kmp_search("test", "this is a test"), true);
        assert_eq!(kmp_search("not", "this is a test"), false);
        assert_eq!(kmp_search("", "any string"), true);
        assert_eq!(kmp_search("abc", "abc"), true);
        assert_eq!(kmp_search("abc", "ab"), false);
    }

    #[test]
    fn test_lcs() {
        assert_eq!(lcs("AGGTAB", "GXTXAYB"), 4);
        assert_eq!(lcs("ABC", "ABC"), 3);
        assert_eq!(lcs("ABC", "DEF"), 0);
        assert_eq!(lcs("", "ABC"), 0);
        assert_eq!(lcs("ABC", ""), 0);
    }
}